src/hotspot/share/gc/shenandoah/shenandoahHeap.inline.hpp

```diff
@@ -132,28 +132,28 @@ inline void ShenandoahHeap::conc_update_with_forwarded(T* p) {
       // Either we succeed in updating the reference, or something else gets in our way.
       // We don't care if that is another concurrent GC update, or another mutator update.
       // We only check that non-NULL store still updated with non-forwarded reference.
-      oop witness = cas_oop(fwd, p, obj);
+      oop witness = atomic_update_oop(fwd, p, obj);
       shenandoah_assert_not_forwarded_except(p, witness, (witness == NULL) || (witness == obj));
     }
   }
 }
 
-inline oop ShenandoahHeap::cas_oop(oop n, oop* addr, oop c) {
+inline oop ShenandoahHeap::atomic_update_oop(oop n, oop* addr, oop c) {
   assert(is_aligned(addr, HeapWordSize), "Address should be aligned: " PTR_FORMAT, p2i(addr));
-  return (oop) Atomic::cmpxchg(addr, c, n);
+  return (oop) Atomic::cmpxchg(addr, c, n, memory_order_relaxed);
 }
 
-inline oop ShenandoahHeap::cas_oop(oop n, narrowOop* addr, narrowOop c) {
+inline oop ShenandoahHeap::atomic_update_oop(oop n, narrowOop* addr, narrowOop c) {
   assert(is_aligned(addr, sizeof(narrowOop)), "Address should be aligned: " PTR_FORMAT, p2i(addr));
   narrowOop val = CompressedOops::encode(n);
-  return CompressedOops::decode(Atomic::cmpxchg(addr, c, val));
+  return CompressedOops::decode(Atomic::cmpxchg(addr, c, val, memory_order_relaxed));
```

Comment:
Are you sure it is sufficient? I would think it needs acq/rel pair, otherwise, read side can see incomplete oop ...

Reply:
Actually, I think you are right: we must ensure the cumulativity of the barriers. Let me think a bit more about it.

Reply:
I think I convinced myself there is a need for `memory_order_acq_rel`. I added a sketch of (counter-)example in code comments. See if that what you were concerned about?

Reply:
Actually, what I meant is that CAS here should use memory_order_release and load barrier needs memory_order_acquire.
I am not sure memory_order_acq_rel is sufficient, C++ states:

A read-modify-write operation with this memory order is both an acquire operation and a release operation. No memory reads or writes in the current thread can be reordered before or after this store. All writes in other threads that release the same atomic variable are visible before the modification and **the modification is visible in other threads that acquire the same atomic variable.**

Reply:
Actually, what I meant is that CAS here should use memory_order_release and load barrier needs memory_order_acquire.
I am not sure memory_order_acq_rel is sufficient, C++ states _memory_order_acq_rel_:

A read-modify-write operation with this memory order is both an acquire operation and a release operation. No memory reads or writes in the current thread can be reordered before or after this store. All writes in other threads that release the same atomic variable are visible before the modification and **the modification is visible in other threads that acquire the same atomic variable.**

Reply:
Ah, I missed JDK-8261495, which is counterpart of this change.  I think memory_order_release is good.